mapper/collectd_mapper/src/functional_batcher.rs

```diff
@@ -0,0 +1,345 @@
+//! Batching algorithm that is unaware of IO. Imperative shell, functional core.
+
+use crate::collectd::OwnedCollectdMessage;
+use clock::Timestamp;
+
+#[derive(Debug, PartialEq)]
+pub struct MessageBatch(pub Vec<OwnedCollectdMessage>);
+
+/// The Batcher's internal state.
+pub struct Batcher {
+    /// Maximum number of messages per batch.
+    max_batch_size: usize,
+
+    /// The maximum age of a batch.
+    ///
+    /// Age of a batch is the elapsed time since `current_batch.opened_at`.
+    max_batch_age: chrono::Duration,
+
+    /// We start a new batch upon receiving a message whose timestamp is farther away to the
+    /// timestamp of first messsge in the batch than `collectd_timestamp_delta` seconds.
+    ///
+    /// Delta is inclusive.
+    collectd_timestamp_delta: f64,
+
+    current_batch: CurrentBatch,
+}
+
+// Invariants:
+//
+//   opened_at.is_none() => messages.is_empty().
+//   opened_at.is_some() => messages.len() > 0.
+//
+struct CurrentBatch {
+    opened_at: Option<Timestamp>,
+    messages: Vec<OwnedCollectdMessage>,
+}
+
+impl CurrentBatch {
+    fn empty() -> Self {
+        Self {
+            opened_at: None,
+            messages: Vec::new(),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.messages.is_empty()
+    }
+}
+
+/// Inputs to the `Batcher`.
+pub enum Input {
+    /// A message was received.
+    Message {
+        /// Time the message has been received.
+        received_at: Timestamp,
+
+        /// The message itself.
+        message: OwnedCollectdMessage,
+    },
+
+    /// Time has progressed.
+    ///
+    /// Allows the Batcher to close a batch when the current time window has expired.
+    Tick {
+        /// The current system time.
+        now: Timestamp,
+    },
+
+    /// Will flush the current batch. This is used upon termination.
+    Flush,
+}
+
+/// Outputs from the `Batcher`. These inform the imperative shell to perform some actions on behalf
+/// of the `Batcher`.
+#[derive(Debug, PartialEq)]
+pub enum Output {
+    /// Informs the imperative shell to send an `Input::Tick` at (or slightly after) the specified timestamp.
+    NextTickAt(Timestamp),
+
+    /// Informs the imperative shell to send the message batch out.
+    MessageBatch(MessageBatch),
+}
+
+impl Batcher {
+    pub fn new(
+        max_batch_size: usize,
+        max_batch_age: chrono::Duration,
+        collectd_timestamp_delta: f64,
+    ) -> Self {
+        assert!(max_batch_size > 0);
+        assert!(collectd_timestamp_delta >= 0.0);
+        Self {
+            max_batch_size,
+            max_batch_age,
+            collectd_timestamp_delta,
+            current_batch: CurrentBatch::empty(),
+        }
+    }
+
+    pub fn handle(&mut self, input: Input, outputs: &mut Vec<Output>) {
```

Comment:
Why not returning a vector rather than populated one provided by the caller?

Reply:
It's an optimization. You can reuse the same `Vec` for every call.

Reply:
Fully agree that in a functional language I'd have returned a `Vec`.